#1.**Principal Component Analysis**

In [ ]:
#Importing the required libraries

import pandas as pd
import numpy as np
from sklearn.decomposition import PCA

import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.linear_model import LinearRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.cluster import KMeans
from sklearn import metrics

from sklearn.preprocessing import StandardScaler
from sklearn import svm

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

In [3]:
#Reading the dataset
df_cc = pd.read_csv("Dataset/CC.csv")
df_cc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8950 entries, 0 to 8949
Data columns (total 18 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   CUST_ID                           8950 non-null   object 
 1   BALANCE                           8950 non-null   float64
 2   BALANCE_FREQUENCY                 8950 non-null   float64
 3   PURCHASES                         8950 non-null   float64
 4   ONEOFF_PURCHASES                  8950 non-null   float64
 5   INSTALLMENTS_PURCHASES            8950 non-null   float64
 6   CASH_ADVANCE                      8950 non-null   float64
 7   PURCHASES_FREQUENCY               8950 non-null   float64
 8   ONEOFF_PURCHASES_FREQUENCY        8950 non-null   float64
 9   PURCHASES_INSTALLMENTS_FREQUENCY  8950 non-null   float64
 10  CASH_ADVANCE_FREQUENCY            8950 non-null   float64
 11  CASH_ADVANCE_TRX                  8950 non-null   int64  
 12  PURCHA

In [4]:
#Having a look on the dataset statistics
df_cc.describe()

BALANCE  BALANCE_FREQUENCY     PURCHASES  ONEOFF_PURCHASES  \
count   8950.000000        8950.000000   8950.000000       8950.000000   
mean    1564.474828           0.877271   1003.204834        592.437371   
std     2081.531879           0.236904   2136.634782       1659.887917   
min        0.000000           0.000000      0.000000          0.000000   
25%      128.281915           0.888889     39.635000          0.000000   
50%      873.385231           1.000000    361.280000         38.000000   
75%     2054.140036           1.000000   1110.130000        577.405000   
max    19043.138560           1.000000  49039.570000      40761.250000   

       INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
count             8950.000000   8950.000000          8950.000000   
mean               411.067645    978.871112             0.490351   
std                904.338115   2097.163877             0.401371   
min                  0.000000      0.000000             0.000000   
25%                  0.000000      0.000000             0.083333   
50%                 89.000000      0.000000             0.500000   
75%                468.637500   1113.821139             0.916667   
max              22500.000000  47137.211760             1.000000   

       ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
count                 8950.000000                       8950.000000   
mean                     0.202458                          0.364437   
std                      0.298336                          0.397448   
min                      0.000000                          0.000000   
25%                      0.000000                          0.000000   
50%                      0.083333                          0.166667   
75%                      0.300000                          0.750000   
max                      1.000000                          1.000000   

       CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
count             8950.000000       8950.000000    8950.000000   8949.000000   
mean                 0.135144          3.248827      14.709832   4494.449450   
std                  0.200121          6.824647      24.857649   3638.815725   
min                  0.000000          0.000000       0.000000     50.000000   
25%                  0.000000          0.000000       1.000000   1600.000000   
50%                  0.000000          0.000000       7.000000   3000.000000   
75%                  0.222222          4.000000      17.000000   6500.000000   
max                  1.500000        123.000000     358.000000  30000.000000   

           PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT       TENURE  
count   8950.000000       8637.000000       8950.000000  8950.000000  
mean    1733.143852        864.206542          0.153715    11.517318  
std     2895.063757       2372.446607          0.292499     1.338331  
min        0.000000          0.019163          0.000000     6.000000  
25%      383.276166        169.123707          0.000000    12.000000  
50%      856.901546        312.343947          0.000000    12.000000  
75%     1901.134317        825.485459          0.142857    12.000000  
max    50721.483360      76406.207520          1.000000    12.000000

In [5]:
#Viewing all the columnns in the dataset.
df_cc[:]

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0     C10001    40.900749           0.818182      95.40              0.00   
1     C10002  3202.467416           0.909091       0.00              0.00   
2     C10003  2495.148862           1.000000     773.17            773.17   
3     C10004  1666.670542           0.636364    1499.00           1499.00   
4     C10005   817.714335           1.000000      16.00             16.00   
...      ...          ...                ...        ...               ...   
8945  C19186    28.493517           1.000000     291.12              0.00   
8946  C19187    19.183215           1.000000     300.00              0.00   
8947  C19188    23.398673           0.833333     144.40              0.00   
8948  C19189    13.457564           0.833333       0.00              0.00   
8949  C19190   372.708075           0.666667    1093.25           1093.25   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                      95.40      0.000000             0.166667   
1                       0.00   6442.945483             0.000000   
2                       0.00      0.000000             1.000000   
3                       0.00    205.788017             0.083333   
4                       0.00      0.000000             0.083333   
...                      ...           ...                  ...   
8945                  291.12      0.000000             1.000000   
8946                  300.00      0.000000             1.000000   
8947                  144.40      0.000000             0.833333   
8948                    0.00     36.558778             0.000000   
8949                    0.00    127.040008             0.666667   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                       0.000000                          0.083333   
1                       0.000000                          0.000000   
2                       1.000000                          0.000000   
3                       0.083333                          0.000000   
4                       0.083333                          0.000000   
...                          ...                               ...   
8945                    0.000000                          0.833333   
8946                    0.000000                          0.833333   
8947                    0.000000                          0.666667   
8948                    0.000000                          0.000000   
8949                    0.666667                          0.000000   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                   0.000000                 0              2        1000.0   
1                   0.250000                 4              0        7000.0   
2                   0.000000                 0             12        7500.0   
3                   0.083333                 1              1        7500.0   
4                   0.000000                 0              1        1200.0   
...                      ...               ...            ...           ...   
8945                0.000000                 0              6        1000.0   
8946                0.000000                 0              6        1000.0   
8947                0.000000                 0              5        1000.0   
8948                0.166667                 2              0         500.0   
8949                0.333333                 2             23        1200.0   

         PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
0      201.802084        139.509787          0.000000      12  
1     4103.032597       1072.340217          0.222222      12  
2      622.066742        627.284787          0.000000      12  
3        0.000000               NaN          0.000000      12  
4      678.334763        244.791237          0.000000      12  
...           ...               ...               ...     ...  
8945   325.594462         48.886365  

In [6]:
#Verifying the columns with NA value.
df_cc.isna().sum()

CUST_ID                               0
BALANCE                               0
BALANCE_FREQUENCY                     0
PURCHASES                             0
ONEOFF_PURCHASES                      0
INSTALLMENTS_PURCHASES                0
CASH_ADVANCE                          0
PURCHASES_FREQUENCY                   0
ONEOFF_PURCHASES_FREQUENCY            0
PURCHASES_INSTALLMENTS_FREQUENCY      0
CASH_ADVANCE_FREQUENCY                0
CASH_ADVANCE_TRX                      0
PURCHASES_TRX                         0
CREDIT_LIMIT                          1
PAYMENTS                              0
MINIMUM_PAYMENTS                    313
PRC_FULL_PAYMENT                      0
TENURE                                0
dtype: int64

In [8]:
#Filling the NA values using mean.
df_cc['MINIMUM_PAYMENTS'].fillna(df_cc['MINIMUM_PAYMENTS'].mean(), inplace=True)
df_cc.dropna(axis=1,inplace=True)

In [9]:
#Dropping the columns which doesn't impact our target.
df_cc.drop(['CUST_ID'], axis=1, inplace=True)

In [10]:
# Applying K-means with 3 clusters
nclusters = 3
km = KMeans(n_clusters=nclusters)
km.fit(df_cc)
print(km)
# Predicting the cluster for each data point.
y_cluster_kmeans = km.predict(df_cc)
score = metrics.silhouette_score(df_cc, y_cluster_kmeans)
print(score)

/usr/local/lib/python3.9/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3)
0.5502197877782317


####a. Applying PCA on CC dataset.

In [12]:
# Using Principal Component Analysis
pca = PCA(2)
x_pca = pca.fit_transform(df_cc)
df2 = pd.DataFrame(data=x_pca)  # Redifining our dataset after for PCA applied dataset

#### b. Apply k-means algorithm on the PCA result and report your observation if the silhouette score hasimproved or not?

In [13]:
# K-means with 3 clusters with PCA
nclusters = 3
km = KMeans(n_clusters=nclusters)
km.fit(df2)
print(km)
# Predicting the cluster for each data point.
y_cluster_kmeans = km.predict(df2)
score = metrics.silhouette_score(df2, y_cluster_kmeans)
print(score)

/usr/local/lib/python3.9/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3)
0.6047436023261318


**Observation:** We see that the Silhoutte_score has improved from 55.02% to 60.47%. Hence there is an increase of ~5.0% in the Silhoutte_score after applying PCA.

#### c.Perform Scaling+PCA+K-Means and report performance.

In [14]:
# Scaling of Features
scaler = StandardScaler()
scaled_df = scaler.fit_transform(df2)

In [15]:
# K-means with 3 clusters with PCA and scaled features
nclusters = 3
km = KMeans(n_clusters=nclusters)
km.fit(scaled_df)
print(km)
# Predicting the cluster for each data point
y_cluster_kmeans = km.predict(scaled_df)
score = metrics.silhouette_score(scaled_df, y_cluster_kmeans)
print(score)

/usr/local/lib/python3.9/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3)
0.6339379378657516


**Reporting Performance:** Here we observe that the Silhoutte_score has improved from 60.47% to 63.39%. Hence there is an increase of ~3.0% in the Sihoutte score after scaling the features using standard scaler.

#2. **Use pd_speech_features.csv**

In [18]:
#Reading the file
speech_df = pd.read_csv('Dataset/pd_speech_features.csv')

In [19]:
#Getting the information/details of the file
speech_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Columns: 755 entries, id to class
dtypes: float64(749), int64(6)
memory usage: 4.4 MB


In [20]:
#Verifying the class distribution.
speech_df['class'].value_counts()

1    564
0    192
Name: class, dtype: int64

In [21]:
#Defining the data model
X = speech_df.drop(['class'],axis=1)
y = speech_df['class']
X.columns

Index(['id', 'gender', 'PPE', 'DFA', 'RPDE', 'numPulses', 'numPeriodsPulses',
       'meanPeriodPulses', 'stdDevPeriodPulses', 'locPctJitter',
       ...
       'tqwt_kurtosisValue_dec_27', 'tqwt_kurtosisValue_dec_28',
       'tqwt_kurtosisValue_dec_29', 'tqwt_kurtosisValue_dec_30',
       'tqwt_kurtosisValue_dec_31', 'tqwt_kurtosisValue_dec_32',
       'tqwt_kurtosisValue_dec_33', 'tqwt_kurtosisValue_dec_34',
       'tqwt_kurtosisValue_dec_35', 'tqwt_kurtosisValue_dec_36'],
      dtype='object', length=754)

####b. Apply PCA (k=3)

In [22]:
# Using Principal Component Analysis
pca = PCA(3)
x_pca = pca.fit_transform(X)
X2 = pd.DataFrame(data=x_pca)

#### c. Perform Scaling

In [23]:
#Scaling the features using StandardScaler()
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [24]:
#Dividing the data into train and test
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25)

#### c. Use SVM to report performance

In [25]:
# SVM with PCA and scaled features
clf = svm.SVC()
clf.fit(X_train, y_train)
clf.score(X_train,y_train)

0.9171075837742504

In [26]:
#Preding the target
y_pred = clf.predict(X_test)

In [27]:
# Accuracy score using SVM with PCA and scaled features
metrics.accuracy_score(y_test, y_pred)

0.8253968253968254

**Reporting Performance:** Here we observe that we achieved an accuracy score of 82.53 using scaling and applying PCA(K=3)

# 3. **Apply Linear Discriminant Analysis (LDA) on Iris.csv dataset to reduce dimensionality of data to k=2.**

In [29]:
#Reading the data
iris_data = pd.read_csv('Dataset/Iris.csv')

In [30]:
#Defining our dataset and target variable
X = iris_data[['SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm']]
y = iris_data['Species']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [31]:
# Using Linear Discriminant Analysis

lda = LinearDiscriminantAnalysis(n_components=2)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)

In [32]:
# Using classification using Random Forest Classifier with LDA
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train, y_train)
classifier.score(X_train, y_train)

0.9821428571428571

In [33]:
#Predicting the target variable.
y_pred = classifier.predict(X_test)

#Viewing the confusion matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)
print()
print('Accuracy : ' + str(accuracy_score(y_test, y_pred)))

[[18  0  0]
 [ 3  7  0]
 [ 0  0 10]]

Accuracy : 0.9210526315789473


# 4. **Briefly identify the difference between PCA and LDA**
###**PCA vs LDA**
Both PCA and LDA are linear transformation techniques. However, PCA is an unsupervised while LDA is a supervised dimensionality reduction technique.

###***Principal Component Analysis***

PCA summarizes a set of features without relying on the output. PCA attempts to find the indications of greatest variance in  a data set. A large set of functions has many functions that are simply copies of  other functions or are closely related to other functions. Such features are essentially unnecessary and can be ignored. The task of PCA is to find such highly correlated or overlapping features and create a new  set of features with minimal correlation between features, in other words, a set of features where the feature difference is as large as possible. Since the variance between  features does not depend on the output,  PCA does not consider the names of the outputs.

###***Linear Discriminant Analysis***

LDA attempts to reduce the dimensionality of the feature set while preserving the information that separates the result classes. LDA tries to find a decision boundary around each cluster of the class. It then projects the data points into the new dimensions so that the clusters are as far apart as possible and the individual elements of the cluster are as close as possible to the center of gravity of the cluster. The new measures are classified by their ability to maximize the distance between  clusters and minimize the distance between cluster data points  and their centroids. These new dimensions form the linear separators of the feature set.